Add docs for torch.compile(numpy) (#109710)
pytorch/pytorch: docs/source/torch.compiler_faq.rst

       return Z

   X = np.random.randn(1000, 1000)
   Y = np.random.randn(1000, 1000)
   with torch.device("cuda"):
       Z = numpy_fn(X, Y)

In this example, ``numpy_fn`` will be executed in CUDA. For this to be
possible, ``torch.compile`` automatically moves ``X`` and ``Y`` from CPU
to CUDA, and then it moves the result ``Z`` from CUDA to CPU. If we are
executing this function several times in the same program run, we may want
to avoid all these rather expensive memory copies. To do so, we just need
to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:

.. code-block:: python

   @torch.compile
   def numpy_fn(X: Tensor, Y: Tensor): -> Tensor
       X = X.numpy()
       Y = X.numpy()
       # Compute Z here
       Z = torch.from_numpy(Z)
       return Z

   X = torch.randn(1000, 1000, device="cuda")
   Y = torch.randn(1000, 1000, device="cuda")
   with torch.device("cuda"):
       Z = numpy_fn(X, Y)

By doing this, we explicitly create the tensors in CUDA memory, and we keep
them there. Note that the original program would not run on eager mode now.
If you want to run it in eager mode, you would need to call ``.numpy(force=True)``

> and perhaps doing ``Z = Z.cuda()`` before returning ``Z``. Of course, doing

Reviewer: Actually, I start wondering what `.numpy(force=True)` does under `torch.compile`

Author: See https://github.com/pytorch/pytorch/pull/109636

Reviewer: of course :-).

contains a few tips and tricks on how to tackle this task.

If the above is not enough to pinpoint the origin of the issue, there are still
a few other NumPy specific tools we can use. If we find a bug that is blocking
our development and we are tracing through mixed PyTorch <> NumPy code where
the NumPy part is not particularly bulky, we can simply deactivate the NumPy
tracing altogether by doing

.. code-block:: python

   from torch._dynamo import config config.trace_numpy = False

This moves back to the behavior in 2.0 and will avoid tracing through any NumPy
function.

Of course, this is not a satisfactory answer if our program is mostly composed
of NumPy code. In these cases, we can try to execute eagerly (without
``torch.compile``) the NumPy code on PyTorch by importing ``import torch._numpy
as np``. This should just be used for **debugging purposes** and is in no way a
replacement for the PyTorch API, as it is **much less performant** and, as a
private API, **may change without notice**. At any rate, ``torch._numpy`` is a

> Python implementation of NumPy in terms of PyTorchand it is used internally to

Reviewer: ```suggestion
Python implementation of NumPy in terms of PyTorch and it is used internally to
```

> transform NumPy code into Pytorch code.. It is rather easy to read and modify,

Reviewer: ```suggestion
transform NumPy code into Pytorch code. It is rather easy to read and modify,
```

``torch.func.functional_call`` to extract the module state.

Does NumPy work with ``torch.compile``?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Starting in 2.1, ``torch.compile`` understands native NumPy programs that
work on NumPy arrays, and mixed PyTorch-NumPy programs that convert from PyTorch
to NumPy and back via ``x.numpy()``, ``torch.from_numpy``, and related functions.

.. _nonsupported-numpy-feats:

Which NumPy features does ``torch.compile`` support?
----------------------------------------------------

NumPy within ``torch.compile`` follows the latest NumPy release.

Generally, ``torch.compile`` is able to trace through most NumPy constructions,
and when it cannot, it falls back to eager and lets NumPy execute that piece of
code. Even then, there are a few features where ``torch.compile`` semantics
slightly deviate from those of NumPy:

- NumPy scalars: We model them as 0-D arrays. That is, ``np.float32(3)`` returns
  a 0-D array under ``torch.compile``. To avoid a graph break, it is best to use this 0-D
  array. If this is not possible, one may often recover the original behavior
  by casting the NumPy scalar to the relevant Python scalar type ``bool/int/float``.

> - Negative strides: ``np.flip`` and slicing with a negative step return a copy.

Reviewer 2: Consider inverting 
```suggestion
- Negative strides: views that result in negative strides will instead copy, e.g. ``np.flip`` or slicing with a negative step.
```

- Type promotion: NumPy's type promotion will change in NumPy 2.0. The new rules
  are described in `NEP 50 <https://numpy.org/neps/nep-0050-scalar-promotion.html)>`__.
  ``torch.compile`` implements NEP 50 rather than the current soon-to-be deprecated rules.

> - ``{tril,triu}_indices_from`` return arrays rather than lists of tuples.

Reviewer 2: If I'm reading the NumPy docs right, this should be
"lists of tuples" -> "a tuple of arrays"

What about other functions that return tuple of array, e.g `numpy.nonzero`. Or for that matter `tril_indices` without the `_from`.

Author: yep, we were missing those. Added.

Author: (nonzero we call it with `as_tuple=True`) I don't worry about that one too much, given that it's data-dependent.

Reviewer 2: What about nonzero?

Author: see the comment above.

There are other features for which we do not support tracing and we gracefully
fallback to NumPy for their execution:

- Non-numeric dtypes like datetimes, strings, chars, void, structured dtypes and recarrays.

- Long dtypes ``np.float128/np.complex256`` and some unsigned dtypes ``np.uint16/np.uint32/np.uint64``.

- ``ndarray`` subclasses.

- Masked arrays.

- Esoteric ufunc machinery like ``axes=[(n,k),(k,m)->(n,m)]`` and ufunc methods (e.g., ``np.add.reduce``).

- Sorting / ordering ``complex64/complex128`` arrays.

- NumPy ``np.poly1d`` and ``np.polynomial``.

- Positional ``out1, out2`` args in functions with 2 or more returns (``out=tuple`` does work).

- ``__array_function__``, ``__array_interface__`` and ``__array_wrap__``.

> - ``ndarray.ctypes`` attribute not supported.

Reviewer 2: Redundant, this is a list of unsupported things. Or does this comment mean that we don't support falling back to eager?
```suggestion
- ``ndarray.ctypes`` attribute.
```

Can I execute NumPy code on CUDA via ``torch.compile``?
-------------------------------------------------------

> Yes you can! To do so, you may simply execute your code under ``torch.device("cuda")``

Reviewer 2: ```suggestion
Yes you can! To do so, you may simply execute your code within a ``torch.device("cuda")`` context
```

.. code-block:: python

   @torch.compile
   def numpy_fn(X: ndarray, Y: ndarray): -> ndarray
       # Compute the ndarray Z here

>        return Z

Reviewer 2: Would be better to give a runnable example so the reader can copy-paste and see it running.

them there. In this case ``X.numpy()`` and ``from_numpy()`` are hints to the compiler
but no real data movement happens. Note that the original program would not run
on eager mode now. If you want to run it in eager mode, you would need to call
``.numpy(force=True)`` doing ``Z = Z.cuda()`` before returning
``Z``. Of course, doing this would execute the program on eager mode NumPy, and
on CPU.

How do I debug my ``torch.compile``d NumPy code?
------------------------------------------------

Debugging JIT compiled code is challenging, given the complexity of modern
compilers and the daunting errors that they raise.
`The tutorial on how to diagnose runtime errors within torch.compile <https://pytorch.org/docs/main/torch.compiler_troubleshooting.html#diagnosing-runtime-errors>`__
contains a few tips and tricks on how to tackle this task.

If the above is not enough to pinpoint the origin of the issue, there are still
a few other NumPy specific tools we can use. If we find a bug that is blocking
our development and we are tracing through mixed PyTorch <> NumPy code where
the NumPy part is not particularly bulky, we can simply deactivate the NumPy

> tracing altogether by doing

Reviewer 2: This reads to me more like a workaround than a debugging method. Perhaps this should mention that this allows us to isolate the bug if it's entirely in the pytorch code?

Reviewer 2: You haven't updated this paragraph.

Author: I updated the paragraph after the code to reflect the point you made.

Reviewer 2: >If we find a bug that is blocking our development [...] we can simply deactivate the NumPy tracing altogether by doing

This is explicitly advising it as a workaround, rather than a debugging method. I'm not a fan of starting the paragraph with one narrative and changing it entirely in the last sentence.

Author: what about now?

.. code-block:: python

>    from torch._dynamo import config config.trace_numpy = False

Reviewer 2: ```suggestion
   from torch._dynamo import config
   config.trace_numpy = False
```

> This moves back to the behavior in 2.0 and will avoid tracing through any NumPy
> function.

Reviewer 2: The reader might not be familiar with the exact 2.0 behavior, so I don't think this comment is helpful.
```suggestion
This disables tracing through any NumPy functions.
```

In fact that is also redundant with the paragraph above, so it might be better to just drop this sentence entirely.

Of course, this is not a satisfactory answer if our program is mostly composed

> of NumPy code. In these cases, we can try to execute eagerly (without

Reviewer 2: "our program is mostly composed of NumPy code" -> "the bug is in the traced NumPy code"

> ``torch.compile``) the NumPy code on PyTorch by importing ``import torch._numpy as np``.

Reviewer 2: Just a thought, but this may be a useful tool in and of itself. If you could enable a flag and dynamo will not inline from `torch._numpy` but will do the conversion from numpy to `torch._numpy`. That way you don't have to edit any code at all to do this test.

Author: The point of changing the imports is that, in most cases, you can execute the code without using dynamo at all. Then, if you see the same bug while doing this, the bug is certainly in the decomposition.

I found this *very* helpful when debugging issues, as when doing this you are dealing with very simple Python code and debugging stuff is as easy as it gets.

This should just be used for **debugging purposes** and is in no way a
replacement for the PyTorch API, as it is **much less performant** and, as a
private API, **may change without notice**. At any rate, ``torch._numpy`` is a
Python implementation of NumPy in terms of PyTorch and it is used internally to
transform NumPy code into Pytorch code. It is rather easy to read and modify,

> so if you find any bug in it feel free to submit a PR fixing it!

Reviewer 2: Could be read as "PRs are okay but issues aren't". Should probably clarify.

If the program does work when importing ``torch._numpy as np``, chances are
that the bug is in TorchDynamo. If this is the case, please feel open an issue

> with a minimal reproducer.

Reviewer 2: Link to docs on running the minifier?

Author: I don't know whether the minifier works with NumPy, but sure.

Reviewer 2: That sounds worth checking. If it doesn't then open an issue?

Author: I agree, but let me punt on this for now and do it later in the week, as I just don't have the bandwidth now.

I ``torch.compile``d a NumPy function and I did not see any speed-up.
--------------------------------------------------------------------

The best place to start is the

> `tutorial with general advice for how to debug this sort of torch.compile issues <https://pytorch.org/docs/main/torch.compiler_faq.html#why-am-i-not-seeing-speedups>`__.

Reviewer 2: ```suggestion
`tutorial with general advice for how to debug these sort of torch.compile issues <https://pytorch.org/docs/main/torch.compiler_faq.html#why-am-i-not-seeing-speedups>`__.
```

Some graph breaks may happen because of the use of unsupported features. See
:ref:`nonsupported-numpy-feats`. More generally, it is useful to keep in mind
that some widely used NumPy features do not play well with compilers. For
example, in-place modifications make reasoning difficult, so the compiler
removes them in a pass called "functionalization". As such, it is best to avoid
in-place ops, or the use of the ``out=`` parameter, and instead simply use

> out-of-place ops and let ``torch.compile`` optimize the memory use. Same goes

Reviewer 2: The messaging is a bit confused here. If the compiler can functionalize automatically, why do I have to change my code?

parameters of the ``nn.Module``. To get this to work, use
``torch.func.functional_call`` to extract the module state.

Does NumPy work with ``torch.compile``?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Starting in 2.1, ``torch.compile`` understands native NumPy programs that
work on NumPy arrays, and mixed PyTorch-NumPy programs that convert from PyTorch
to NumPy and back via ``x.numpy()``, ``torch.from_numpy``, and related functions.

.. _nonsupported-numpy-feats:

Which NumPy features does ``torch.compile`` support?
----------------------------------------------------

> NumPy within ``torch.compile`` follows the latest NumPy release.

Reviewer 3: do we have some way of getting the version of NumPy that torch.compile works / is tested against?

Author: Nope. This is more of something we would like to have, and a rule for people to submit issues / for us to answer issues.

Reviewer 3: cool, can you file an issue to follow-up on this?

Author: I think there is no issue to follow up on here. I think we should say that we support the last NumPy release, or I can change it to "the same NumPy release that PyTorch supports". I don't think we have any way to get this, but I remember it's written somewhere. @albanD knows where IIRC. I can change the wording and put a link to the relevant place where we track this.

Reviewer 3: What I mean is the following:
- 1. clearly we are encoding some NumPy behavior based on some version of NumPy, and as written it seems like this can differ from the _installed_ version of NumPy.
- 2. "Latest" as a descriptor isn't super helpful for the user - they could be using an old version of PyTorch or NumPy could have just updated between PyTorch releases.
- 3.  If we know what 1. is, we should just tell the user (in code!) so they can check and do what they will with that information, such as installing a NumPy version to match or upgrading PyTorch.

Author: For iii, do we currently have a way to retrieve the NumPy version we support when using it in eager or when running the tests?

Reviewer 4: @lezcano isn't the north star story here be that PyTorch runtime should match the behavior of the currently installed numpy version (with obviously only a subset of versions being properly supported and some leading to nice unsupported errors).
As of today we're basically targetting a given version of Numpy 2.X right? And I would agree that we should make sure that we have CI running with it.

Author: Yes, at the moment we are implementing the last version of NumPy (pretty much, modulo that NEP50 point). Then I guess we'll be able to add support for NumPy 2.0 (while keeping support for 1.X as well) once we support it in core.

Reviewer 4: That sounds good. I would just not say "the last version of NumPy" but "the Numpy 2.0 pre-release" so that it doesn't become stale if we don't update it.

Author: That makes complete sense. Updating

Generally, ``torch.compile`` is able to trace through most NumPy constructions,
and when it cannot, it falls back to eager and lets NumPy execute that piece of
code. Even then, there are a few features where ``torch.compile`` semantics
slightly deviate from those of NumPy:

- NumPy scalars: We model them as 0-D arrays. That is, ``np.float32(3)`` returns
  a 0-D array under ``torch.compile``. To avoid a graph break, it is best to use this 0-D
  array. If this is not possible, one may often recover the original behavior
  by casting the NumPy scalar to the relevant Python scalar type ``bool/int/float``.

> - Negative strides: ``np.flip`` and slicing with a negative step return a copy.

Author: We could, but we found that the pattern `x[::-1].fn()` where `fn` is an out-of-place op is relatively common (it's in torchbench even) so I think it's worth this small dissonance. If people don't agree, we can always reconsider and fall back, or even consider implementing this ourselves, as inductor does support negative strides.

-------------------------------------------------------

Yes you can! To do so, you may simply execute your code within a ``torch.device("cuda")``
context. Consider the example

.. code-block:: python

   import torch
   import numpy as np

   @torch.compile
   def numpy_fn(X: np.ndarray, Y: np.ndarray) -> np.ndarray:
       return np.sum(X[:, :, None] * Y[:, None, :], axis=(-2, -1))

   X = np.random.randn(1024, 64)
   Y = np.random.randn(1024, 64)
   with torch.device("cuda"):
       Z = numpy_fn(X, Y)

In this example, ``numpy_fn`` will be executed in CUDA. For this to be
possible, ``torch.compile`` automatically moves ``X`` and ``Y`` from CPU
to CUDA, and then it moves the result ``Z`` from CUDA to CPU. If we are
executing this function several times in the same program run, we may want
to avoid all these rather expensive memory copies. To do so, we just need
to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:

.. code-block:: python

   @torch.compile
   def numpy_fn(X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
       X, Y = X.numpy(), Y.numpy()
       Z = np.sum(X[:, :, None] * Y[:, None, :], axis=(-2, -1))
       return torch.from_numpy(Z)

>    X = torch.randn(1024, 64)
>    Y = torch.randn(1024, 64)

Reviewer 2: Copy-paste error:
```suggestion
   X = torch.randn(1024, 64, device="cuda")
   Y = torch.randn(1024, 64, device="cuda")
```

``.numpy(force=True)`` doing ``Z = Z.cuda()`` before returning
``Z``. Of course, doing this would execute the program on eager mode NumPy, and
on CPU.

How do I debug my ``torch.compile``d NumPy code?
------------------------------------------------

Debugging JIT compiled code is challenging, given the complexity of modern
compilers and the daunting errors that they raise.
`The tutorial on how to diagnose runtime errors within torch.compile <https://pytorch.org/docs/main/torch.compiler_troubleshooting.html#diagnosing-runtime-errors>`__
contains a few tips and tricks on how to tackle this task.

If the above is not enough to pinpoint the origin of the issue, there are still
a few other NumPy specific tools we can use. If we find a bug that is blocking
our development and we are tracing through mixed PyTorch <> NumPy code where
the NumPy part is not particularly bulky, we can simply deactivate the NumPy
tracing altogether by doing

.. code-block:: python

   from torch._dynamo import config config.trace_numpy = False

This moves back to the behavior in 2.0 and will avoid tracing through any NumPy
function.

> Of course, this is not a satisfactory answer if our program is mostly composed

Reviewer 3: I didn't really follow this sentence -- why is it not satisfactory?  I think you mean that you want to debug it rather than just turning it off.  In that case, I'd put the whole thought in the first sentence.

Author: Makes sense. I'll rewrite it.

of NumPy code. In these cases, we can try to execute eagerly (without
``torch.compile``) the NumPy code on PyTorch by importing ``import torch._numpy as np``.
This should just be used for **debugging purposes** and is in no way a
replacement for the PyTorch API, as it is **much less performant** and, as a
private API, **may change without notice**. At any rate, ``torch._numpy`` is a
Python implementation of NumPy in terms of PyTorch and it is used internally to
transform NumPy code into Pytorch code. It is rather easy to read and modify,
so if you find any bug in it feel free to submit a PR fixing it!

If the program does work when importing ``torch._numpy as np``, chances are
that the bug is in TorchDynamo. If this is the case, please feel open an issue
with a minimal reproducer.

I ``torch.compile`d a NumPy function and I did not see any speed-up.
--------------------------------------------------------------------

The best place to start is the
`tutorial with general advice for how to debug this sort of torch.compile issues <https://pytorch.org/docs/main/torch.compiler_faq.html#why-am-i-not-seeing-speedups>`__.

> Some graph breaks may happen because of the use of unsupported features. See

Author: The intuition behind this (and something that I tried to follow both in post and the docs) is that the reader may be someone that comes from NumPy and does not necessarily have too much PyTorch experience. As such, I tried to crossref quite a bit and avoid assuming too much PyTorch knowledge.

  a 0-D array under ``torch.compile``. To avoid a graph break, it is best to use this 0-D
  array. If this is not possible, one may often recover the original behavior
  by casting the NumPy scalar to the relevant Python scalar type ``bool/int/float``.

- Negative strides: ``np.flip`` and slicing with a negative step return a copy.

- Type promotion: NumPy's type promotion will change in NumPy 2.0. The new rules
  are described in `NEP 50 <https://numpy.org/neps/nep-0050-scalar-promotion.html)>`__.
  ``torch.compile`` implements NEP 50 rather than the current soon-to-be deprecated rules.

- ``{tril,triu}_indices_from`` return arrays rather than lists of tuples.

There are other features for which we do not support tracing and we gracefully
fallback to NumPy for their execution:

- Non-numeric dtypes like datetimes, strings, chars, void, structured dtypes and recarrays.

- Long dtypes ``np.float128/np.complex256`` and some unsigned dtypes ``np.uint16/np.uint32/np.uint64`.

- ``ndarray`` subclasses.

- Masked arrays.

- Esoteric ufunc machinery like ``axes=[(n,k),(k,m)->(n,m)]`` and ufunc methods (e.g., ``np.add.reduce``).

> - Fortran ordered arrays and, in general, any ``order=`` different to ``C``.

Reviewer: (sorry, did not catch it on the first read). 
Strictly speaking, some functions have defaults different from 'C' ('A' or 'K'). This does not matter in practice, and we graph break / raise NotImplementedError for non-default order arguments. 

Maybe just say any non-default values of the `order=` arguments?

Author: I'm just going to remove this line as we also graphbreak on a few other inputs and it'd be a pain to list them all.
